### PR TITLE
Migrate Using Airflow to execute SQL

### PR DESCRIPTION
**Pull request by jwitz**

### CODE REVIEW

File: guides/airflow-sql.md [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+```sql
+-- call-sproc2
+CALL sp_pi_squared();
+```
+
+`sp_pi()` and `sp_pi_squared()` are two stored procedures that are defined in a Snowflake instance. Note that the SQL in these files could be any type of query you need to execute. Sprocs are used here only as an example.
+
+Finally, you need to set up a connection to Snowflake. There are a few ways to manage connections using Astronomer, including [IAM roles](https://docs.astronomer.io/software/integrate-iam), [secrets managers](https://docs.astronomer.io/software/secrets-backend), and the [Airflow API](https://docs.astronomer.io/software/airflow-api). For this example, set up a connection using the Airflow UI. Because the connection in the DAG is called `snowflake`, your configured connection should look something like this:
+
+![Configured Snowflake connection in the Airflow UI](img/guides/snowflake_connection.png)
```
Comment (jwitz) [inline guides/airflow-sql.md:144]:
```suggestion
![Configured Snowflake connection in the Airflow UI](/img/guides/snowflake_connection.png)
```

File: guides/airflow-sql.md [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+            file_format='covid_csv',
+            snowflake_conn_id='snowflake'
+        )
+
+        t0 >> generate_files >> snowflake
+```
+
+Here's a graph view of the DAG:
+
+![Covid-to-Snowflake Graph](img/guides/covid_to_snowflake_graph_view.png)
```
Comment (jwitz) [inline guides/airflow-sql.md:302]:
```suggestion
![Covid-to-Snowflake Graph](/img/guides/covid_to_snowflake_graph_view.png)
```

File: guides/airflow-sql.md [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+- A defined Snowflake stage (`covid_stage`) and file format (`covid_csv`). See the [Snowflake documentation](https://docs.snowflake.com/en/user-guide/data-load-s3.html).
+
+Next, set up your Airflow connections. This example requires two connections:
+
+- A connection to S3 (established using `astro-s3-workshop` in the DAG above).
+- A connection to Snowflake (established using `snowflake`. See Example 1 for a screenshot of what the connection should look like).
+
+After this setup, you're ready to run the DAG! After a successful run, you can see the new data for today's date in your table.
+
+![Snowflake Data](img/guides/snowflake_data_populated.png)
```
Comment (jwitz) [inline guides/airflow-sql.md:316]:
```suggestion
![Snowflake Data](/img/guides/snowflake_data_populated.png)
```

File: guides/airflow-sql.md [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+
+dag_factory = dagfactory.DagFactory("/usr/local/airflow/include/config_file.yml")
+
+dag_factory.clean_dags(globals())
+dag_factory.generate_dags(globals())
+```
+
+Once you Deploy it, the DAG will show a single task the Airflow UI:
+
+![DAG Factory Graph](img/guides/dag_factory_graph_view.png)
```
Comment (jwitz) [inline guides/airflow-sql.md:437]:
```suggestion
![DAG Factory Graph](/img/guides/dag_factory_graph_view.png)
```

File: guides/airflow-sql.md [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+
+```sql
+SELECT *
+FROM STATE_DATA
+WHERE date = {{ yesterday_ds_nodash }}f
+```
+
+In this example, the query has been parameterized to dynamically select data for yesterday's date using a built-in Airflow variable with double curly brackets. The rendered template in the Airflow UI looks like this:
+
+![Rendered Template](img/guides/rendered_template.png)
```
Comment (jwitz) [inline guides/airflow-sql.md:194]:
```suggestion
![Rendered Template](/img/guides/rendered_template.png)
```

File: guides/airflow-sql.md [hunk 6]
```diff
@@ -0,0 +1,447 @@
+---
+title: "Using Airflow to Execute SQL"
+description: "Execute queries, parameterize queries, and embed SQL-driven ETL in Apache Airflow DAGs."
+id: airflow-sql
+sidebar_label: "Execute SQL queries"
+---
+
+Executing SQL queries is one of the most common use cases for data pipelines. Whether you're extracting and loading data, calling a stored procedure, or executing a complex query for a report, Airflow has you covered. Using Airflow, you can orchestrate all of your SQL tasks elegantly with just a few lines of boilerplate code.
+
+In this guide, we'll cover general best practices for executing SQL from your DAG, showcase Airflow's available SQL-related operators, and demonstrate how to use Airflow for a few common SQL use cases.
```
Comment (kwcanuck) [inline guides/airflow-sql.md:10]:
```suggestion
In this guide you'll learn about the best practices for executing SQL from your DAG,  review the  most commonly used Airflow SQL-related operators, and then use sample code to implement a few common SQL use cases.
```

File: guides/airflow-sql.md [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+
+:::info
+
+All code in this guide can be found on the [Astronomer GitHub](https://github.com/astronomer/airflow-sql-tutorial).
```
Comment (kwcanuck) [inline guides/airflow-sql.md:14]:
```suggestion
All code used in this tutorial is located in the [Astronomer Registry](https://github.com/astronomer/airflow-sql-tutorial).
```

File: guides/airflow-sql.md [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+- [S3ToSnowflakeTransferOperator](https://registry.astronomer.io/providers/snowflake/modules/s3tosnowflakeoperator)
+- [S3toRedshiftOperator](https://registry.astronomer.io/providers/amazon/modules/s3toredshiftoperator)
+- [GCSToBigQueryOperator](https://registry.astronomer.io/providers/google/modules/gcstobigqueryoperator)
+- [PostgresToGCSOperator](https://registry.astronomer.io/providers/google/modules/postgrestogcsoperator)
+- [BaseSQLToGCSOperator](https://registry.astronomer.io/providers/google/modules/basesqltogcsoperator)
+- [VerticaToMySqlOperator](https://registry.astronomer.io/providers/mysql/modules/verticatomysqloperator)
+
+## Examples
+
+With those basic concepts in mind, we'll show a few examples of common SQL use cases. For this tutorial you'll use [Snowflake](https://www.snowflake.com/), but note that the concepts shown can be adapted for other databases. Some of the environment setup for each example makes use of the [Astro CLI](https://docs.astronomer.io/astro/cli/overview) and Astro project structure, but you can also adapt this setup for use with Apache Airflow.
```
Comment (kwcanuck) [inline guides/airflow-sql.md:83]:
```suggestion
Now that you've learned about the most commonly used Airflow SQL operators, you'll use the operators in some SQL use cases. For this tutorial you'll use [Snowflake](https://www.snowflake.com/), but the concepts shown can be adapted for other databases. Some of the environment setup for each example makes use of the [Astro CLI](https://docs.astronomer.io/astro/cli/overview) and Astro project structure, but you can also adapt this setup for use with Apache Airflow.
```

File: guides/airflow-sql.md [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+
+:::info
+
+All code in this guide can be found on the [Astronomer GitHub](https://github.com/astronomer/airflow-sql-tutorial).
+
+:::
+
+## Best practices for executing SQL from your DAG
+
+No matter which database and flavor of SQL you're using, there are many ways to execute  your queries using Airflow. Once you determine how to execute your queries, the following tips will help you keep your DAGs clean, readable, and efficient for execution.
```
Comment (kwcanuck) [inline guides/airflow-sql.md:20]:
```suggestion
No matter what database or SQL version you're using, there are many ways to execute your queries using Airflow. Once you determine how to execute your queries, the following tips will help you keep your DAGs clean, readable, and efficient for execution.
```

File: guides/airflow-sql.md [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+
+### Use hooks and operators
+
+Using hooks and operators whenever possible makes your DAGs easier to read, easier to maintain, and more performant. Airflow has many SQL-related operators available that can significantly limit the code needed to execute your queries.
```
Comment (kwcanuck) [inline guides/airflow-sql.md:24]:
```suggestion
Using hooks and operators whenever possible makes your DAGs easier to read, easier to maintain, and improves performance. The SQL-related operators included with Airflow can significantly limit the code needed to execute your queries.
```

File: guides/airflow-sql.md [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+
+An exception to this rule could be very short queries (such as `SELECT * FROM table`). Putting one-line queries like this directly in the DAG is fine if it makes your code more readable.
+
+### Keep transformations in SQL
+
+Remember that Airflow is primarily an orchestrator, not a transformation framework. While you have the full power of Python in your DAG, Astronomer recommends offloading as much of your transformation logic as possible to third party transformation frameworks. With SQL, this means completing the transformations within your query whenever possible.
+
+## SQL operators
+
+Airflow has many operators available out of the box that make working with SQL easier. This guide lists some of the most commonly used ones that you should be aware of, but note that this list isn't comprehensive. For more documentation about Airflow operators, head [here](https://airflow.apache.org/docs/stable/_api/airflow/operators/index.html#).
```
Comment (kwcanuck) [inline guides/airflow-sql.md:52]:
```suggestion
To make working with SQL easier, Airflow includes many built in operators. This tutorial discusses some of the most commonly used operators and shouldn't be considered a definitive resource. For more information about the available Airflow operators, see [airflow.operators](https://airflow.apache.org/docs/stable/_api/airflow/operators/index.html#).
```

File: guides/airflow-sql.md [hunk 12]
```diff
@@ -0,0 +1,447 @@
...
+         ) as dag:
+
+         opr_param_query = SnowflakeOperator(
+             task_id='param_query',
+             snowflake_conn_id='snowflake',
+             sql='param-query.sql'
+         )
+```
+
+The DAG is essentially the same as in Example 1. The difference is in the query itself:
```
Comment (kwcanuck) [inline guides/airflow-sql.md:184]:
```suggestion
The DAG is essentially the same that you used in Example 1. The difference is in the query itself:
```

File: guides/airflow-sql.md [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+SELECT *
+FROM STATE_DATA
+WHERE date = {{ yesterday_ds_nodash }}f
+```
+
+In this example, the query has been parameterized to dynamically select data for yesterday's date using a built-in Airflow variable with double curly brackets. The rendered template in the Airflow UI looks like this:
+
+![Rendered Template](/img/guides/rendered_template.png)
+
+Astronomer recommends using Airflow variables or macros whenever possible to increase flexibility and make your workflows [idempotent](https://en.wikipedia.org/wiki/Idempotence). The above example will work with any Airflow variables. For example, you could access a variable from your Airflow config:
```
Comment (kwcanuck) [inline guides/airflow-sql.md:196]:
```suggestion
Astronomer recommends using Airflow variables or macros whenever possible to increase flexibility and make your workflows [idempotent](https://en.wikipedia.org/wiki/Idempotence). The above example will work with any Airflow variables. For example, you could access a variable from your Airflow config:
```

" increase flexibility"?

File: guides/airflow-sql.md [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+- [GCSToBigQueryOperator](https://registry.astronomer.io/providers/google/modules/gcstobigqueryoperator)
+- [PostgresToGCSOperator](https://registry.astronomer.io/providers/google/modules/postgrestogcsoperator)
+- [BaseSQLToGCSOperator](https://registry.astronomer.io/providers/google/modules/basesqltogcsoperator)
+- [VerticaToMySqlOperator](https://registry.astronomer.io/providers/mysql/modules/verticatomysqloperator)
+
+## Examples
+
+With those basic concepts in mind, we'll show a few examples of common SQL use cases. For this tutorial you'll use [Snowflake](https://www.snowflake.com/), but note that the concepts shown can be adapted for other databases. Some of the environment setup for each example makes use of the [Astro CLI](https://docs.astronomer.io/astro/cli/overview) and Astro project structure, but you can also adapt this setup for use with Apache Airflow.
+
+### Example 1: Executing a query
```
Comment (kwcanuck) [inline guides/airflow-sql.md:85]:
```suggestion
### Example 1: Execute a query
```

File: guides/airflow-sql.md [hunk 15]
```diff
@@ -0,0 +1,447 @@
...
+
+`sp_pi()` and `sp_pi_squared()` are two stored procedures that are defined in a Snowflake instance. Note that the SQL in these files could be any type of query you need to execute. Sprocs are used here only as an example.
+
+Finally, you need to set up a connection to Snowflake. There are a few ways to manage connections using Astronomer, including [IAM roles](https://docs.astronomer.io/software/integrate-iam), [secrets managers](https://docs.astronomer.io/software/secrets-backend), and the [Airflow API](https://docs.astronomer.io/software/airflow-api). For this example, set up a connection using the Airflow UI. Because the connection in the DAG is called `snowflake`, your configured connection should look something like this:
+
+![Configured Snowflake connection in the Airflow UI](/img/guides/snowflake_connection.png)
+
+With the connection established, you can now run the DAG to execute the SQL queries.
+
+### Example 2: Executing a query with parameters
```
Comment (kwcanuck) [inline guides/airflow-sql.md:148]:
```suggestion
### Example 2: Execute a query with parameters
```

File: guides/airflow-sql.md [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+
+Astronomer recommends using Airflow variables or macros whenever possible to increase flexibility and make your workflows [idempotent](https://en.wikipedia.org/wiki/Idempotence). The above example will work with any Airflow variables. For example, you could access a variable from your Airflow config:
+
+```sql
+SELECT *
+FROM STATE_DATA
+WHERE state = {{ conf['state_variable'] }}
+```
+
+If you need a parameter that is not available as a built-in variable/macro, such as a value from another task in your DAG, you can also pass that parameter into your query using the operator:
```
Comment (kwcanuck) [inline guides/airflow-sql.md:204]:
```suggestion
If you need a parameter that is not available as a built-in variable or a macro, such as a value from another task in your DAG, you can also pass that parameter into your query using the operator:
```

File: guides/airflow-sql.md [hunk 17]
```diff
@@ -0,0 +1,447 @@
...
+
+And then reference that param in your SQL file:
+
+```sql
+SELECT *
+FROM STATE_DATA
+WHERE date = {{ params.date }}
+```
+
+### Example 3: Loading data
```
Comment (kwcanuck) [inline guides/airflow-sql.md:223]:
```suggestion
### Example 3: Load data
```

File: guides/airflow-sql.md [hunk 18]
```diff
@@ -0,0 +1,447 @@
...
+
+The next example loads data from an external source into a table in a database. You'll grab data from an API and save it to a flat file on S3, which you can then load into Snowflake.
+
+This example uses the [S3toSnowflakeTransferOperator](https://registry.astronomer.io/providers/snowflake/modules/s3tosnowflakeoperator) to limit the code that you have to write.
+
+First, create a DAG that pulls COVID data from an [API endpoint](https://covidtracking.com/data/api) for California, Colorado, Washington, and Oregon, saves the data to comma-separated values (CSVs) on S3, and loads each of those CSVs to Snowflake using the transfer operator. Here's the DAG code:
```
Comment (kwcanuck) [inline guides/airflow-sql.md:229]:
```suggestion
First, create a DAG that pulls COVID data from an [API endpoint](https://covidtracking.com/data/api) for California, Colorado, Washington, and Oregon, saves the data to comma-separated values (CSVs) on Amazon S3, and loads each of those CSVs to Snowflake using the transfer operator. Here's the DAG code:
```

File: guides/airflow-sql.md [hunk 19]
```diff
@@ -0,0 +1,447 @@
...
+            table='STATE_DATA',
+            schema='SANDBOX_KENTEND',
+            file_format='covid_csv',
+            snowflake_conn_id='snowflake'
+        )
+
+        t0 >> generate_files >> snowflake
+```
+
+Here's a graph view of the DAG:
```
Comment (kwcanuck) [inline guides/airflow-sql.md:300]:
```suggestion
This image shows a graph view of the DAG:
```

File: guides/airflow-sql.md [hunk 20]
```diff
@@ -0,0 +1,447 @@
...
+- A connection to S3 (established using `astro-s3-workshop` in the DAG above).
+- A connection to Snowflake (established using `snowflake`. See Example 1 for a screenshot of what the connection should look like).
+
+After this setup, you're ready to run the DAG! After a successful run, you can see the new data for today's date in your table.
+
+![Snowflake Data](/img/guides/snowflake_data_populated.png)
+
+Note that while this example is specific to Snowflake, the concepts apply to any database you might be using. If a transfer operator doesn't exist for your specific source and destination tools, you can always write your own (and maybe contribute it back to the Airflow project)!
+
+## Example 4: Using pandas
```
Comment (kwcanuck) [inline guides/airflow-sql.md:320]:
```suggestion
## Example 4: Use pandas
```

File: guides/airflow-sql.md [hunk 21]
```diff
@@ -0,0 +1,447 @@
...
+
+While Astronomer recommends using SQL-related operators and keeping any data transformations in SQL, for some use cases this doesn't work. For instance, pivoting data into a new format for a report can be difficult to complete with SQL alone. In this next example, you'll make use of Python libraries to integrate your SQL into a Python function.
```
Comment (kwcanuck) [inline guides/airflow-sql.md:322]:
```suggestion
While Astronomer recommends using SQL-related operators and keeping any data transformations in SQL, for some use cases this doesn't work. For instance, pivoting data into a new format for a report can be difficult to complete with SQL alone. In this next example, you'll make use of Python libraries to integrate your SQL operator into a Python function.
```

File: guides/airflow-sql.md [hunk 22]
```diff
@@ -0,0 +1,447 @@
...
+            table='PIVOT_STATE_DATA',
+            schema='SANDBOX_KENTEND',
+            file_format='covid_csv',
+            snowflake_conn_id='snowflake'
+        )
+
+        opr_pivot_data >> opr_load_data
+```
+
+In the DAG, the Python function `pivot_data` executes the SQL query and saves the results in a pandas dataframe using the `read_sql` function. It then pivots the data to the desired format and saves the it to S3. Lastly, the downstream task `opr_load_data`  loads that data back to Snowflake using the transfer operator described in Example 3.
```
Comment (kwcanuck) [inline guides/airflow-sql.md:395]:
```suggestion
In the DAG, the Python function `pivot_data` executes the SQL query and saves the results in a pandas dataframe using the `read_sql` function. It then pivots the data to the desired format and saves it to Amazon S3. Lastly, the downstream task `opr_load_data`  loads that data back to Snowflake using the transfer operator described in Example 3.
```

File: guides/airflow-sql.md [hunk 23]
```diff
@@ -0,0 +1,447 @@
...
+
+## Example 5 - Using dag-factory
```
Comment (kwcanuck) [inline guides/airflow-sql.md:397]:
```suggestion
## Example 5 - Use dag-factory
```

File: guides/airflow-sql.md [hunk 24]
```diff
@@ -0,0 +1,447 @@
...
+
+If you have SQL users who aren't familiar with Airflow or don't know any Python, they can use [dag-factory](https://github.com/ajbosco/dag-factory) to generate DAGs using a YAML configuration file.
```
Comment (kwcanuck) [inline guides/airflow-sql.md:399]:
```suggestion
If you're unfamiliar with Airflow or Python, you can use [dag-factory](https://github.com/ajbosco/dag-factory) to generate DAGs using a YAML configuration file.
```

File: guides/airflow-sql.md [hunk 25]
```diff
@@ -0,0 +1,447 @@
...
+dag_factory.generate_dags(globals())
+```
+
+Once you Deploy it, the DAG will show a single task the Airflow UI:
+
+![DAG Factory Graph](/img/guides/dag_factory_graph_view.png)
+
+## Next steps
+
+This guide covered how to interact with your SQL database from Airflow and important best practices when doing so. But there are still some outstanding questions:
```
Comment (kwcanuck) [inline guides/airflow-sql.md:441]:
```suggestion
You've learned how to interact with your SQL database from Airflow. But there are still some outstanding questions:
```

**Review by kwcanuck — APPROVED** [review]

**Comment by kwcanuck** [comment]
@jwitz it doesn't appear that you've added the tutorial  to docs/sidebarsGuides.js.

File: guides/airflow-sql.md (hunk 7 above)
Comment (jwitz) [reply guides/airflow-sql.md:14]: I don't think this link is on the registry yet - updating comment

**Pull request merged**